File: IT/L'Indirizzario/Python/Indirizzario Python v3/MainFrame.py
# file MainFrame.py

try:
	from tkinter import Frame, Menu
except ImportError:
	from Tkinter import Frame, Menu

from Pmw import NoteBook
from Contact import * 
from NContact import *
import sqlite3

class MainFrame(Frame):
	def __init__(self):
		Frame.__init__(self, height=480, width=640, master=None)

		# crea il Notebook

		self.tabbedPane = NoteBook(self)

		self.conn = sqlite3.connect("addrb.db")
		cursor = self.conn.cursor()

		cursor.execute("CREATE TABLE IF NOT EXISTS contacts (id INTEGER PRIMARY KEY, name VARCHAR, surname VARCHAR, addr VARCHAR, pnum VARCHAR, email VARCHAR, CHECK(LENGTH(pnum <= 15)))")

		# salva (eventuali) modifiche al database

		self.conn.commit()

		self.updateUI()

		# crea un nuovo menu

		menubar = Menu(self)
		fileMenu = Menu(menubar)
		menubar.add_cascade(label="file", menu=fileMenu)

		# aggiunge tre sottomenu al menu "file", associando a ciascuno il relativo gestore degli eventi

		fileMenu.add_command(label="Nuovo", command=self.save)
		fileMenu.add_command(label="Cancella", command=self.remove)
		fileMenu.add_command(label="Esci", command=self.exit)

		# imposta la barra dei menu

		self.master.config(menu=menubar)

		self.tabbedPane.pack(fill='both', expand=True)

		self.pack_propagate(False)
		self.pack()
		self.master.wm_title("Indirizzario")

	# il metodo save() gestisce il click sul menu "Nuovo"

	def save(self):
		NContact(self)

	# il metodo remove() gestisce il click sul menu "Cancella"

	def remove(self):
		pageName = self.tabbedPane.getcurselection()
		index = self.tabbedPane.index(pageName)
		tab = self.tabbedPane.page(index)

		# recupera il pannello Contact associato alla linguetta con il "monkey patching" 

		contact = tab.contact

		name = contact.name.__getitem__("text")
		surname = contact.surname.__getitem__("text")

		# cancella il contatto dal database

		cursor = self.conn.cursor()
		cursor.execute("DELETE FROM contacts WHERE name = ? AND surname = ?", (name, surname))

		# salva le modifiche sul database

		self.conn.commit()
		self.updateUI()

	# il metodo exit() gestisce il click sul menu "Esci"

	def exit(self): 
		self.master.destroy()

	# crea le linguette dei contatti nel Notebook

	def updateUI(self):
		self.deleteTabs()
		cursor = self.conn.cursor()
		cursor.execute("SELECT name, surname, addr, pnum, email FROM contacts")

		i = 1
		for row in cursor:
			pageName = "{} - {} {}".format(i, row[0], row[1])
			tab = self.tabbedPane.add(pageName)
			contact = Contact(parent=tab, name=row[0], surname=row[1], addr=row[2], pnum=row[3], email=row[4]) 

			# utilizza il "monkey patching" per aggiungere un contatto alla linguetta

			tab.contact = contact
			i += 1

	# cancella tutte le linguette nel NoteBook

	def deleteTabs(self):
		for page in self.tabbedPane.pagenames():
			self.tabbedPane.delete(page)

File: IT/2/L'Indirizzario/C++/Indirizzario C++ v3/MainFrame.cpp
/* file MainFrame.cpp */

#include <QApplication>
#include <QMenuBar>
#include <QVBoxLayout>
#include <QSqlDatabase>
#include <QSqlQuery>
#include "MainFrame.h"
#include "Contact.h"
#include "NContact.h"

namespace ind = programmiamo::ind;

ind::MainFrame::MainFrame(QWidget *parent) : QMainWindow(parent) 
{
	/* crea il Notebook */

	tabs = new QTabWidget(this);

	db = QSqlDatabase::addDatabase("QSQLITE");
	db.setDatabaseName("addrb.sqlite");
	db.open();

	/* se non esiste, crea la tabella dei contatti */

	QSqlQuery query("CREATE TABLE IF NOT EXISTS contacts (id INTEGER PRIMARY KEY, name VARCHAR, surname VARCHAR, addr VARCHAR, pnum VARCHAR, email VARCHAR, CHECK(LENGTH(pnum <= 15)))");

	updateUI();

	QMenu *fileMenu = menuBar()->addMenu("&File");

	QAction *saveAction = new QAction("Nuovo", this);
	QAction *removeAction = new QAction("Cancella", this);
	QAction *exitAction = new QAction("Esci", this);

	fileMenu->addAction(saveAction);
	fileMenu->addAction(removeAction);
	fileMenu->addAction(exitAction);

	/* associa a ogni voce del menu il relativo gestore di evento */

	connect(saveAction, SIGNAL(triggered()), this, SLOT(save()));
	connect(removeAction, SIGNAL(triggered()), this, SLOT(remove()));
	connect(exitAction, SIGNAL(triggered()), this, SLOT(exit()));

	QWidget *panel = new QWidget(this);
	this->setCentralWidget(panel);

	QVBoxLayout *box = new QVBoxLayout();
	box->setMargin(0);
	box->addWidget(tabs);
	panel->setLayout(box);

	this->resize(640, 480);
	this->show();
}

/* gestisce il click sul menu "Nuovo", e apre una nuova finestra per inserire un nuovo contatto */

void ind::MainFrame::save()
{
	NContact* window = new NContact(this);
	window->exec();
}

/* gestisce il click sul menu "Cancella", cancellando la linguetta corrente */ 

void ind::MainFrame::remove()
{
	int index = tabs->currentIndex();

	if (index == -1) {
		return;
	}

	QWidget* widget = tabs->currentWidget();
	Contact* tab = dynamic_cast<Contact*>(widget);

	if (tab) {
		QString name = tab->name->text();
		QString surname = tab->surname->text();

		QSqlQuery pStmt;
		pStmt.prepare("DELETE FROM contacts WHERE name = :name AND surname = :surname");
		pStmt.bindValue(":name", name);
		pStmt.bindValue(":surname", surname);
		pStmt.exec();
	}

	tabs->removeTab(index);
}

/* gestisce il click sul menu "Esci", chiudendo l'applicazione */ 

void ind::MainFrame::exit()
{
	QApplication::quit();
}

/* crea le linguette dei contatti nel Notebook */

void ind::MainFrame::updateUI()
{
	tabs->clear();

	QSqlQuery query("SELECT name, surname, addr, pnum, email FROM contacts");

	/* recupera i dati (nome, cognome, ...) dal database */

	while (query.next()) {
		QString name = query.value(0).toString();
		QString surname = query.value(1).toString();
		QString addr = query.value(2).toString();
		QString pnum = query.value(3).toString();
		QString email = query.value(4).toString();

		QString pageName = name + " " + surname;

		Contact *contact = new Contact(tabs, name, surname, addr, pnum, email);
		tabs->addTab(contact, pageName);
	}
}

